File: C-Plus-Plus/cryptography/HillCipher.cpp
/* This code is an implementation of Hill Cipher. It
   is a polygraphic substitution cipher. It forms a 
   nxn matrix of key and nx1 matrix(s) of the plaintext
   and then multiplies the two matrices to obtains the
   matrix of ciphertext.*/

#include<bits/stdc++.h>
using namespace std;
#define SIZE 3

//This function prints the cipher text
void printciphertext(int cipherMatrix[SIZE][1]){
    
    string output;
    for(int i=0;i<SIZE;i++){
        output += cipherMatrix[i][0]+97;
    }

    cout<<"The cipher text is :"<<output;
}

//This function calculates the matrix multiplication
void cipher(int cipherMatrix[SIZE][1],int keyMatrix[SIZE][SIZE],int plaintextMatrix[SIZE][1]){

      //calculating the multiplication 
      for(int i=0;i<SIZE;i++){
          cipherMatrix[i][0] = 0;
          for(int j=0;j<SIZE;j++){
              cipherMatrix[i][0] += keyMatrix[i][j]*plaintextMatrix[j][0];
          }

          cipherMatrix[i][0] = cipherMatrix[i][0]%26;
      }     
}

//This function builds the key matrix
void buildkeyMatrix(int keyMatrix[SIZE][SIZE],string key){
 
    int counter =0;
    //storing the key in matrix
    for(int i=0;i<SIZE;i++){
        for(int j=0;j<SIZE;j++){
            keyMatrix[i][j] = key[counter]%97;
            counter++; 
        }
    }
}

//This function builds the plaintext matrix
void buildplaintextMatrix(int plaintextMatrix[][1],string plaintext){

    //storing the plaintext in matrix form
    for(int i=0;i<SIZE;i++){
        plaintextMatrix[i][0] = (plaintext[i])%97;
    }
}

//This fuinction helps us in calling all the functions in sequential order
void hillcipher(string plaintext,string key){

    int plaintextMatrix[SIZE][1];
    buildplaintextMatrix(plaintextMatrix,plaintext);

    int keyMatrix[SIZE][SIZE];
    buildkeyMatrix(keyMatrix,key);

    int cipherMatrix[SIZE][1];
    cipher(cipherMatrix,keyMatrix,plaintextMatrix);

    printciphertext(cipherMatrix);
}

int main(){

    string plaintext;
    string key;

    cout<<"The plaintext size you need to enter is :"<<SIZE<<endl;

    cout<<"Enter the plaintext here same :";
    //Taking plaintext from here
    cin>>plaintext;
   
    cout<<"Enter the key here :";
    //Taking key from here
    cin>>key; 

    int size1 = plaintext.length();
    int size2 = key.length(); 

    //validating for the input given 
    if(size2 != size1*size1 || size1 != SIZE){
        cout<<"The key input is wrong!"<<endl;
        return 0;
    }
    else 
        hillcipher(plaintext,key);

    return 0;
}
/*
    Sample I/O :

    The plaintext size you need to enter is :3
    Enter the plaintext here same :ask
    Enter the key here :thisiskey
    The cipher text is :yma
*/

File: Python/cryptography/HillCipher.py
'''This code is an implementation of Hill Cipher. It
   is a polygraphic substitution cipher. It forms a 
   nxn matrix of key and nx1 matrix(s) of the plaintext
   and then multiplies the two matrices to obtains the
   matrix of ciphertext.'''

import sys

#This function builds key-Martix given the key
def buildKeyMatrix(key):
   
   k = 0
   for i in range(n):
     for j in range(n):
       keyMatrix[i][j] = ord(key[k])%97
       k += 1

#This function builds Message-Matrix given planetext   
def buildMessageMatrix(plaintext):
   
   for i in range(n): 
     messageMatrix[i][0] = ord(plaintext[i])%97 

'''This is the encryption function which takes planetext
   and key as input and returns us the ciphertext list.'''   
def encription(plaintext,key):
   
   buildKeyMatrix(key)
   buildMessageMatrix(plaintext)    
   
   #matrix multiplication of key and message matrices
   for i in range(n):
     for j in range(n):
       cipherMatrix[i][0] += ( keyMatrix[i][j]*messageMatrix[j][0] )
     cipherMatrix[i][0]    = cipherMatrix[i][0] %26
   
   #storing all the result as a list
   cipherText = []
   for i in range(n):
     cipherText.append( chr(cipherMatrix[i][0]+97) )

   return cipherText 
     

if __name__=='__main__':
   
   #Taking plaintext and key as input from user
   plaintext = input("Enter the plaintext here :")
   key       = input("Enter the key here :")
   
   n = len(plaintext)
   
   if len(key) != n*n :
      print("Invalid Key")
      sys.exit(0)
   
   #creating the matrices required
   keyMatrix     = [ [0]*n for i in range(n) ]
   messageMatrix = [ [0]   for i in range(n) ]
   cipherMatrix  = [ [0]   for i in range(n) ]
   
   for i in range(n):
      cipherMatrix[i][0] = 0
  
   '''calling the encryption function on the given plaintext
      and key''' 
   encrypted_message = encription(plaintext,key)
   
   #printing the ciphertext we got using the encryption method
   print("Encripted plaintext is :","".join(encrypted_message) )

'''
Sample I/O:

   Enter the plaintext here :ask 
   Enter the key here :thisiskey
   Encripted plaintext is : yma     
'''
 
